File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/AggregationRangeVariableSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class AggregationRangeVariableSyntax : VisualBasicSyntaxNode
	{
		internal VariableNameEqualsSyntax _nameEquals;

		internal AggregationSyntax _aggregation;

		public VariableNameEqualsSyntax NameEquals => GetRedAtZero(ref _nameEquals);

		public AggregationSyntax Aggregation => GetRed(ref _aggregation, 1);

		internal AggregationRangeVariableSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal AggregationRangeVariableSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, VariableNameEqualsSyntax nameEquals, AggregationSyntax aggregation)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AggregationRangeVariableSyntax(kind, errors, annotations, (nameEquals != null) ? ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VariableNameEqualsSyntax)nameEquals.Green) : null, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AggregationSyntax)aggregation.Green), null, 0)
		{
		}

		public AggregationRangeVariableSyntax WithNameEquals(VariableNameEqualsSyntax nameEquals)
		{
			return Update(nameEquals, Aggregation);
		}

		public AggregationRangeVariableSyntax WithAggregation(AggregationSyntax aggregation)
		{
			return Update(NameEquals, aggregation);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				0 => _nameEquals, 
				1 => _aggregation, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				0 => NameEquals, 
				1 => Aggregation, 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitAggregationRangeVariable(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitAggregationRangeVariable(this);
		}

		public AggregationRangeVariableSyntax Update(VariableNameEqualsSyntax nameEquals, AggregationSyntax aggregation)
		{
			if (nameEquals != NameEquals || aggregation != Aggregation)
			{
				AggregationRangeVariableSyntax aggregationRangeVariableSyntax = SyntaxFactory.AggregationRangeVariable(nameEquals, aggregation);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(aggregationRangeVariableSyntax, annotations);
				}
				return aggregationRangeVariableSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/AggregationRangeVariableSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class AggregationRangeVariableSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode
		Friend _nameEquals As Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableNameEqualsSyntax

		Friend _aggregation As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationSyntax

		Public ReadOnly Property Aggregation As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationSyntax)(Me._aggregation, 1)
			End Get
		End Property

		Public ReadOnly Property NameEquals As Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableNameEqualsSyntax
			Get
				Return MyBase.GetRedAtZero(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableNameEqualsSyntax)(Me._nameEquals)
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal nameEquals As Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableNameEqualsSyntax, ByVal aggregation As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AggregationRangeVariableSyntax(kind, errors, annotations, If(nameEquals IsNot Nothing, DirectCast(nameEquals.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VariableNameEqualsSyntax), Nothing), DirectCast(aggregation.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AggregationSyntax)), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitAggregationRangeVariable(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitAggregationRangeVariable(Me)
		End Sub

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Dim num As Integer = i
			If (num = 0) Then
				syntaxNode = Me._nameEquals
			ElseIf (num = 1) Then
				syntaxNode = Me._aggregation
			Else
				syntaxNode = Nothing
			End If
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim nameEquals As SyntaxNode
			Dim num As Integer = i
			If (num = 0) Then
				nameEquals = Me.NameEquals
			ElseIf (num = 1) Then
				nameEquals = Me.Aggregation
			Else
				nameEquals = Nothing
			End If
			Return nameEquals
		End Function

		Public Function Update(ByVal nameEquals As Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableNameEqualsSyntax, ByVal aggregation As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationRangeVariableSyntax
			Dim aggregationRangeVariableSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationRangeVariableSyntax
			If (nameEquals <> Me.NameEquals OrElse aggregation <> Me.Aggregation) Then
				Dim aggregationRangeVariableSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationRangeVariableSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.AggregationRangeVariable(nameEquals, aggregation)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				aggregationRangeVariableSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, aggregationRangeVariableSyntax1, aggregationRangeVariableSyntax1.WithAnnotations(annotations))
			Else
				aggregationRangeVariableSyntax = Me
			End If
			Return aggregationRangeVariableSyntax
		End Function

		Public Function WithAggregation(ByVal aggregation As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationRangeVariableSyntax
			Return Me.Update(Me.NameEquals, aggregation)
		End Function

		Public Function WithNameEquals(ByVal nameEquals As Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableNameEqualsSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.AggregationRangeVariableSyntax
			Return Me.Update(nameEquals, Me.Aggregation)
		End Function
	End Class
End Namespace